### ISSUE TITLE
TypeScript: aggregate errors for file from all projects

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#20185]
_From @OliverJAsh on September 30, 2017 16:20_

I have a project with three folders, two of which are TS projects and the third of which is shared between them:

- shared
- service-worker (with `tsconfig.json`)
- browser (with `tsconfig.json`)

Both "browser" and "service-worker" depend on "shared".

However, when I open a file in "shared", I only see errors related to one of the `tsconfig.json`s—usually this will be the `tsconfig.json ` used for the file I navigated from previously.

It would be great if VSCode could aggregate all errors for the current file for all matching TS projects.

_Copied from original issue: Microsoft/vscode#35432_

**Comment by mjbvz** [comment]
Related to #20184 as both involve the same file that is part of multiple `tsconfig.json` projects. Again, it is not clear to me what the expected experience would be in this case

**Comment by OliverJAsh** [comment]
Another issue I just hit because of this: "find references" will not show all actual references.

Perhaps the errors and references should be grouped by the `tsconfig.json` path (when there are multiple projects)

**Comment by mhegazy** [comment]
> Another issue I just hit because of this: "find references" will not show all actual references.

not sure i understand this comment, can you elaborate?

**Comment by OliverJAsh** [comment]
When I have a file that is used in multiple TS projects, "find references" will not show references across all of those TS projects.

Steps to reproduce:

1. create 3 TS projects: "browser", "service-worker", "shared"
2. create a module in the "shared" project with an export
3. depend on that module from "browser"/"service-worker"
4. navigate to the module created in "shared"
5. use find references on the export in the "shared" module

**Comment by OliverJAsh** [comment]
@mhegazy Are there any updates on this? I am just starting another project which has similar requirements: two TS projects with their own dependency graph and some shared dependencies (e.g. `client` and `server`, and some files used by both).

Let me know if you want any more information about the problem.

**Comment by mhegazy** [comment]
That sounds like a bug. I have not had a chance to invistigate though.

**Comment by OliverJAsh** [comment]
This issue is really bothering me so here are some clearer steps to reproduce.

We have two TS projects, `project1` and `project2`. (A real world use case would be `server` and `client` aka `frontend` and `backend`.) These TS projects depend on some common shared files.

```
.
├── project1
│   ├── index.ts
│   └── tsconfig.json
├── project2
│   ├── index.ts
│   └── tsconfig.json
└── shared
    └── index.ts
```

``` ts
// project1/index.ts
import { foo } from "../shared/index";
```

``` json
// project1/tsconfig.json
{
    "files": [
        "./index.ts"
    ]
}
```

``` ts
// project2/index.ts
import { foo } from "../shared/index";
```

``` json
// project2/tsconfig.json
{
    "files": [
        "./index.ts"
    ]
}
```

``` ts
// shared/index.ts
export const foo = 1;
```

Finding references on a shared symbol in one of the TS projects only shows references in the TS project of the selected file—VSCode/TypeScript does not aggregate the references of all TS projects that apply to the current file.

E.g. missing references in `project1`:

![image](https://user-images.githubusercontent.com/921609/40162349-a621c2e0-59ab-11e8-8201-876f2a300d79.png)

Finding references on a symbol in the shared file shows no references.

E.g. missing references in `project1` and `project2`:

![image](https://user-images.githubusercontent.com/921609/40162604-6b12d206-59ac-11e8-8888-740b43e82424.png)

… unless VSCode/TypeScript has also opened a file from one of the other TS projects, in which case only references from that TS project will be shown.

E.g. missing references in `project2`:

![image](https://user-images.githubusercontent.com/921609/40162694-a6c2197e-59ac-11e8-8c17-413f22c00a37.png)

I am only demonstrating the behaviour of references, but the same behaviour extends to error reporting, refactors like renaming, etc.

Ideally VSCode/TypeScript would aggregate all TS projects for the current file.

Alternatively, is there a better way to structure the project?

Potentially related issues:
- https://github.com/Microsoft/TypeScript/issues/21965
- https://github.com/Microsoft/vscode/issues/29419
- https://github.com/Microsoft/vscode/issues/29316
- https://github.com/Microsoft/vscode/issues/3772

**Comment by sheetalkamat** [comment]
When you open shared.ts the context matters, since its not directly referenced in any of the tsconfig file in its directory hierarchy ... tsserver will have the information of project/projects it was opened/used in earlier. If project1 is open, there is no way to know that it would also be referenced in project2 unless it is open and hence the behaviour you see. For any file we only look in its hierarchy for the tsconfig file.

**Comment by sheetalkamat** [comment]
> That sounds like a bug. I have not had a chance to invistigate though.

This is not bug @mhegazy . unless the project is open there is no really default config project for shared file to be able to deduce its being referenced in those other projects whose tsconfig's are not even in directory chain.

**Comment by mhegazy** [comment]
this is about having the file in multiple contexts. we already do something similar for find all refs for instance where we repeat the process for every containing project.

**Comment by OliverJAsh** [comment]
@sheetalkamat That explains the current behaviour.

@mhegazy Do you think there's anything TypeScript/VSCode could offer to make the user experience better here? It's quite common to have multiple TS projects with overlapping dependency graphs, e.g. `server` and `client`. When those "shared" files are viewed, the ideal experience would be for TypeScript/VSCode to aggregate errors from all TS projects that file appears within. Similarly, both projects would be used when refactoring or finding references.

I think it's clear that this is the ideal/intuitive user experience, but what are the challenges involved with getting there?

**Comment by OliverJAsh** [comment]
> unless the project is open there is no really default config project for shared file to be able to deduce its being referenced in those other projects whose tsconfig's are not even in directory chain.

Could VSCode not check whether a file appears in all TS projects (in the root folder)? I presume this has performance challenges though?

**Comment by mhegazy** [comment]
> Could TypeScript/VSCode not check whether a file appears in all TS projects (in the root folder)? I presume this has performance challenges though?

yes. the server only knows about open files/projects. a file can be referenced from any where on disk, and it is not manageable to search for possible incoming references everywhere.

I think a better solution for your scenario is https://github.com/Microsoft/TypeScript/issues/3469

**Comment by OliverJAsh** [comment]
> I think a better solution for your scenario is #3469

How do you imagine this (composite projects?) helping the situation? I guess we would have two TS projects, `client` and `server`—and then a composite project that depends on both of them? Would that then mean VSCode/TypeScript would show aggregate all errors (across potentially many TS projects) for a given file?

**Comment by mhegazy** [comment]
you would have a tsconfig.josn at the root that is empty, but has a set of project references to both projects, the server would load that one if it happened to be on the root of your project directory (the folder you used to open vscode `code <folder>`), and that will force **all** your projects to be loaded in memory.
this is still work in progress, so that is not available now.

**Comment by OliverJAsh** [comment]
That makes sense! In that case, I'll close this in favour of #3469

**Comment by OliverJAsh** [comment]
Fixed by #3469

**Comment by OliverJAsh** [comment]
@mhegazy 

Re. https://github.com/microsoft/TypeScript/issues/20185#issuecomment-390425304

> you would have a tsconfig.josn at the root that is empty, but has a set of project references to both projects, the server would load that one if it happened to be on the root of your project directory (the folder you used to open vscode `code <folder>`), and that will force **all** your projects to be loaded in memory.
> this is still work in progress, so that is not available now.

This idea works with `tsc --build`, but I'm having a lot of teething issues getting it to play nicely in VS Code. Is this a problem with VS Code or TS server?

https://github.com/microsoft/vscode/issues/80438
https://github.com/microsoft/vscode/issues/80423

**Comment by RyanCavanaugh** [comment]
Closing language service bugs related to the 6.0 implementation. For more information, see #62827
